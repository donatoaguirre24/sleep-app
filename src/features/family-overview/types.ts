export type SleepIntervalOverview = {
  intervalId: string
  userId: string
  name: string
  averageHeartRate: number
  averageRespiratoryRate: number
  score: number
  timeAsleep: number
}

export type DailyFamilyOverview = {
  date: string
  data: Array<SleepIntervalOverview>
}
