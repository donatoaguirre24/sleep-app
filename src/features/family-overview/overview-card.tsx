import { useTheme } from '@shopify/restyle'
import React, { useMemo } from 'react'
import { Pressable, PressableStateCallbackType } from 'react-native'

import { strings } from '@/locales'
import { getTimeFromSeconds } from '@/services'
import { Box, Card, CardProps, CircularProgress, Text, Theme } from '@/ui'
import { SleepIntervalOverview } from './types'

type Props = CardProps & {
  data: SleepIntervalOverview
  onSelection: (data: SleepIntervalOverview) => void
}

export const OverviewCard: React.FC<Props> = ({ data, onSelection, ...props }) => {
  const { colors } = useTheme<Theme>()
  const { hours, minutes } = getTimeFromSeconds(data.timeAsleep)

  const timeAsleep = `${hours}${strings.common.hoursShort} ${minutes}${strings.common.minutesShort} ${strings.common.asleep}`
  const averageBpm = `${data.averageHeartRate}${strings.common.bpm} (${strings.common.averageShort})`

  const scoreColor = useMemo(() => {
    if (data.score < 50) {
      return colors.red
    }
    if (data.score < 75) {
      return colors.orange
    }
    return colors.green
  }, [data.score, colors])

  const handlePress = () => {
    onSelection(data)
  }

  const computeStylesFromState = ({ pressed }: PressableStateCallbackType) => ({
    opacity: pressed ? 0.6 : 1,
    transform: [{ scale: pressed ? 0.98 : 1 }],
  })

  return (
    <Pressable onPress={handlePress} style={computeStylesFromState}>
      <Card
        backgroundColor="card"
        flexDirection="row"
        alignItems="center"
        variant="elevated"
        paddingVertical="m"
        {...props}
      >
        <CircularProgress
          flex={1}
          marginRight="s"
          color={scoreColor}
          percentage={data.score}
          size={65}
          strokeWidth={5}
        />
        <Box flex={3}>
          <Text variant="subheader" marginBottom="s">
            {data.name}
          </Text>
          <Box flexDirection="row">
            <Box flex={1}>
              <Text color="secondaryGray" fontWeight="600">
                {timeAsleep}
              </Text>
            </Box>
            <Box flex={1} paddingLeft="s">
              <Text color="magenta" fontWeight="600">
                â™¥ {averageBpm}
              </Text>
            </Box>
          </Box>
        </Box>
      </Card>
    </Pressable>
  )
}
