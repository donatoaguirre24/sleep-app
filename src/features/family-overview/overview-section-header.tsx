import { BoxProps } from '@shopify/restyle'
import React from 'react'

import { dateStringFromISODate } from '@/services'
import { Box, Text, Theme } from '@/ui'

type Props = BoxProps<Theme> & {
  date: string
}

export const OverviewSectionHeader: React.FC<Props> = ({ date, ...props }) => {
  const formattedDate = dateStringFromISODate(date)

  return (
    <Box bg="background" {...props}>
      <Text variant="subheader">{formattedDate}</Text>
    </Box>
  )
}
