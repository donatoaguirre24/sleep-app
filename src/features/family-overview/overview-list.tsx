import { useTheme } from '@shopify/restyle'
import React from 'react'
import { SectionList } from 'react-native'

import { Theme } from '@/ui'
import { OverviewCard } from './overview-card'
import { OverviewSectionHeader } from './overview-section-header'
import { DailyFamilyOverview, SleepIntervalOverview } from './types'

type Props = {
  data: Array<DailyFamilyOverview>
  onSelection: (data: SleepIntervalOverview) => void
}

export const OverviewList: React.FC<Props> = ({ data, onSelection }) => {
  const { spacing } = useTheme<Theme>()

  return (
    <SectionList
      contentContainerStyle={{ paddingBottom: spacing.m, paddingHorizontal: spacing.m }}
      contentInsetAdjustmentBehavior="automatic"
      keyExtractor={(item, index) => item.userId + index}
      sections={data}
      renderItem={({ item }) => (
        <OverviewCard marginBottom="m" data={item} onSelection={onSelection} />
      )}
      renderSectionHeader={({ section }) => (
        <OverviewSectionHeader paddingVertical="s" rawDate={section.date} />
      )}
    />
  )
}
