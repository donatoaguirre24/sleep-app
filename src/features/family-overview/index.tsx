import React, { useCallback } from 'react'
import { SafeAreaView } from 'react-native'

import data from '@/data/overview.json'
import { NavigationProps, Screens } from '@/navigation'
import { OverviewList } from './overview-list'

export const FamilyOverview: React.FC<NavigationProps> = ({ navigation }) => {
  const { navigate } = navigation

  const showSleepData = useCallback(
    (intervalId: string) => {
      navigate(Screens.SleepDetails, { intervalId })
    },
    [navigate]
  )

  return (
    <SafeAreaView>
      <OverviewList data={data} onSelection={showSleepData} />
    </SafeAreaView>
  )
}
