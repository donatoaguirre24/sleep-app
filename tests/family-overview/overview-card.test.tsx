import { fireEvent, render, screen } from '@testing-library/react-native'
import React from 'react'

import { OverviewCard } from '@/features/family-overview/overview-card'
import { withProviders } from '../utils'

describe('OverviewCard', () => {
  it('renders correctly', () => {
    render(withProviders(<OverviewCard data={mockData} onSelection={jest.fn()} />))

    expect(screen.toJSON()).toMatchSnapshot()
  })

  it('should render the time asleep properly', () => {
    render(withProviders(<OverviewCard data={mockData} onSelection={jest.fn()} />))

    const timeAsleep = screen.getByText('6h 13m asleep')

    expect(timeAsleep).toBeDefined()
  })

  it('should call onSelection when a card is tapped', () => {
    const mockOnSelection = jest.fn()

    render(withProviders(<OverviewCard data={mockData} onSelection={mockOnSelection} />))

    const card = screen.getByText('John Doe')

    fireEvent.press(card)

    expect(mockOnSelection).toBeCalledWith(mockData)
  })
})

const mockData = {
  intervalId: '1489046760',
  userId: '2228b530e055401f81ba37b51ff6f81d',
  name: 'John Doe',
  averageHeartRate: 84,
  averageRespiratoryRate: 17,
  score: 44,
  timeAsleep: 22380,
}
