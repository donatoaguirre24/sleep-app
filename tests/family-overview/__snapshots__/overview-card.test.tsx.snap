// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OverviewCard renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
        "transform": Array [
          Object {
            "scale": 1,
          },
        ],
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "flexDirection": "row",
            "paddingVertical": 16,
          },
          Array [
            Object {
              "borderColor": "#CCCCCC",
              "borderRadius": 8,
              "borderWidth": 0,
              "padding": 8,
              "shadowColor": "#CCCCCC",
              "shadowOffset": Object {
                "height": 3,
                "width": 1,
              },
              "shadowOpacity": 0.5,
              "shadowRadius": 4,
            },
          ],
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flex": 1,
              "marginRight": 8,
            },
          ]
        }
      >
        <RNSVGSvgView
          align="xMidYMid"
          bbHeight="65"
          bbWidth="65"
          focusable={false}
          height={65}
          meetOrSlice={0}
          minX={0}
          minY={0}
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
                "borderWidth": 0,
              },
              Object {
                "flex": 0,
                "height": 65,
                "width": 65,
              },
            ]
          }
          vbHeight={65}
          vbWidth={65}
          width={65}
        >
          <RNSVGGroup>
            <RNSVGCircle
              cx="50%"
              cy="50%"
              fill={
                Object {
                  "type": 0,
                  "value": 0,
                }
              }
              propList={
                Array [
                  "fill",
                  "stroke",
                  "strokeWidth",
                  "strokeOpacity",
                ]
              }
              r="30"
              stroke={
                Object {
                  "type": 0,
                  "value": 4294924103,
                }
              }
              strokeOpacity={0.2}
              strokeWidth="5"
            />
            <RNSVGCircle
              cx="50%"
              cy="50%"
              fill={
                Object {
                  "type": 0,
                  "value": 0,
                }
              }
              matrix={
                Array [
                  6.123233995736766e-17,
                  -1,
                  1,
                  6.123233995736766e-17,
                  0,
                  65,
                ]
              }
              propList={
                Array [
                  "fill",
                  "stroke",
                  "strokeWidth",
                  "strokeDasharray",
                  "strokeLinecap",
                ]
              }
              r="30"
              stroke={
                Object {
                  "type": 0,
                  "value": 4294924103,
                }
              }
              strokeDasharray={
                Array [
                  "82.93804605477052",
                  "105.55751316061705",
                ]
              }
              strokeLinecap={1}
              strokeWidth="5"
            />
          </RNSVGGroup>
        </RNSVGSvgView>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "justifyContent": "center",
              },
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#1C1C1E",
                  "fontSize": 15,
                  "fontWeight": "bold",
                  "lineHeight": 18,
                  "textAlign": "center",
                },
                Object {
                  "color": "#FF5747",
                },
              ]
            }
          >
            44
            %
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "flex": 3,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#1C1C1E",
                "fontSize": 20,
                "fontWeight": "600",
                "lineHeight": 26,
                "marginBottom": 8,
              },
            ]
          }
        >
          John Doe
        </Text>
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "flex": 1,
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#5E5E5E",
                    "fontSize": 12,
                    "fontWeight": "600",
                    "lineHeight": 18,
                  },
                ]
              }
            >
              6h 13m asleep
            </Text>
          </View>
          <View
            style={
              Array [
                Object {
                  "flex": 1,
                  "paddingLeft": 8,
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#FF006A",
                    "fontSize": 12,
                    "fontWeight": "600",
                    "lineHeight": 18,
                  },
                ]
              }
            >
              â™¥ 
              84bpm (avg)
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
